Module 2 Project Exercise

shivaHanchate opened the pull request:
Hi,
please review my Module 2 project exercise.

Many Thanks
Shiva

shivaHanchate committed:
Added config.py to gitignore

shivaHanchate committed:
Added code to Mark as Complete Button to update status to complete

shivaHanchate committed:
Added code to call API to fetch cards and Move cards

shivaHanchate committed:
Added code for the table to show Cards and update status

shivaHanchate committed:
removed extra button

shivaHanchate committed:
Add code to move from Done to ToDo status

shivaHanchate committed:
Add button to move Status from Done to ToDo

shivaHanchate committed:
add class to the code

shivaHanchate committed:
removed flask_config.py and session_items.py files as they are not being used. Added card.py and trello_api.py files

shivaHanchate committed:
refactored code and Trello_Api  class to trello_api.py

shivaHanchate committed:
added Card class to generate Card details

shivaHanchate committed:
refactored code and separated API code and App code

shivaHanchate committed:
removed hard coded list_id's, replaced it with status

shivaHanchate committed:
Add details.py and amended README.md to add information about details.py file

hugh-emerson commented on card.py:
```diff
@@ -0,0 +1,28 @@
+from config import api_key, token, board_id, todo_list_id, doing_list_id, done_list_id
+
+
+class Card:
```
If you have an instance of this class, `my_first_card = Card()` it is not true to say "my_first_card is a card". It's an object that has a function that returns a list of card data. It should be called something like `class CardService`

Also, `api_key`, `token`, `board_id`, `done_list_id` are unused. And actually there's no real reason this needs to be a class except to support having multiple instances of the class for managing different boards.

What the exercise wanted was a class like this:
```
class Card:
    def __init__(self, id, name, status):
        self.id = id
        self.name = name
        self.status = status
```

And then when you are converting Trello data into your list of card details, you do:
```
trello_card.append(Card(id_card, card_name, status))
```

Or if you want it can still contain the logic for parsing Trello data if you make it like this:

```
class Card:
    def __init__(self, trello_item):
        self.id = trello_item['id']
        self.name = trello_item['name']
        if item['idList'] == todo_list_id:
            self.status = "ToDo"
        elif item['idList'] == doing_list_id:
            self.status = "Doing"
        else:
            self.status = "Done"
```
```
cards = []
for item in trello_data:
    cards.append(Card(item))
```

Let me know by Slack/email if you have any questions.

hugh-emerson commented on card.py:
```diff
@@ -0,0 +1,28 @@
+        self.done_list_id = done_list_id
+
+    def card_details(self, data):
+        trello_card = []
```
This variable should be called `trello_cards` because it is a list of cards, not just one card.

hugh-emerson commented on README.md:
```diff
@@ -37,6 +37,8 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
-* There's also a [SECRET_KEY](https://flask.palletsprojects.com/en/1.1.x/config/#SECRET_KEY) variable which is used to encrypt the flask session cookie.
+
+### Important Note
+Rename details.py to config.py and amend API key, token, board_id, todo_list_id, doing_list_id and done_list_id to match your trello credentials.
```
While this works, it would be neater to use environment variables. In particular, next project exercise's instructions will assume you used environment variables.

Add these values to the `.env` file and then access them in python code with `os.getenv('YOUR_VARIABLE_NAME')`

hugh-emerson commented on app.py:
```diff
@@ -1,20 +1,42 @@
+        return render_template('index.html', items=response)
+    data = trello_api.get_board_cards()
+    response = card.card_details(data)
+    print(response)
```
You would normally remove things like print statements before merging.

hugh-emerson commented on app.py:
```diff
@@ -1,20 +1,42 @@
+        trello_api.add_card(new_todo_item)
+        data = trello_api.get_board_cards()
+        response = card.card_details(data)
+        return render_template('index.html', items=response)
```
This should be a separate route. It is totally separate to the lines outside this block.

So it's much neater to split it into two functions, with different `app.route` definitions - an index() function for GET "/", and an additem() function for POST '/' like you had before.

hugh-emerson replied:
It would be better to redirect to index instead of lines 15-17. Reduces code duplication and also improves behaviour of the site - e.g. if the user refreshes their browser after creating a card it would currently try to resubmit the form (creating another card).

You're doing this correctly in your other endpoints e.g. move_item_to_in_progress

hugh-emerson commented on requirements.txt:
```diff
@@ -1,2 +1,39 @@
-Flask==1.1.1
-python-dotenv==0.12.0
+attrs==20.2.0
```
I think you may have installed some packages you don't need

hugh-emerson commented on trello_api.py:
```diff
@@ -0,0 +1,48 @@
+        return response
+
+    def add_card(self, new_todo_item):
+        endpoint = f"https://api.trello.com/1/cards?name={new_todo_item}"
+        response = requests.post(endpoint, params={"key": self.api_key, "token": self.token, "idList": self.todo_list_id}).json()
```
It's confusing to include one query parameter via your `endpoint` string and then include the other 3 via the `params` dictionary.
The end result is the same wherever you put the query parameter, but splitting it up is harder to read and could lead to problems.
A dictionary is slightly easier to read/manage than building a long string, so I'd suggest moving `name` there.

hugh-emerson reviewed (approved):
Looks good 👍 
I've left some comments with suggestions but merge this whenever you're ready.

In particular:
- It would be good to create a class for individual cards as that's what the task asked for, but you have successfully defined and used classes so I'm still passing the module.
- Also you may want to move your config variables into `.env` to make the next project exercise a bit more straightforward.

shivaHanchate committed:
delete details.py as variables have been moved to .env file

shivaHanchate committed:
Amend .env.template to add variables

shivaHanchate committed:
amend to add environment variables note

shivaHanchate committed:
amend code to use variables from .env file and moved name to params

shivaHanchate committed:
renamed card to CardService , removed variables which are not used and made changes to var names

renamed card to CardService , removed variables which are not used and made changes to var names

shivaHanchate committed:
split the function into two, refactored Get and Post methods

shivaHanchate committed:
separate and refactored Get and Post methods

shivaHanchate committed:
Merge remote-tracking branch 'origin/Module-2' into Module-2
